tokio/src/io/util/mem.rs

```diff
@@ -161,19 +181,24 @@ impl Drop for DuplexStream {
     }
 }
 
-// ===== impl Pipe =====
+// ===== impl SimplexStream =====
 
-impl Pipe {
-    fn new(max_buf_size: usize) -> Self {
-        Pipe {
-            buffer: BytesMut::new(),
-            is_closed: false,
-            max_buf_size,
-            read_waker: None,
-            write_waker: None,
-        }
+/// Creates unidirectional buffer that acts like pair of connected single direction sockets.
+///
+/// The `max_buf_size` argument is the maximum amount of bytes that can be
+/// written to a buffer before the it returns `Poll::Pending`.
+#[cfg_attr(docsrs, doc(cfg(feature = "io-util")))]
+pub fn simplex(max_buf_size: usize) -> SimplexStream {
+    SimplexStream {
+        buffer: BytesMut::new(),
+        is_closed: false,
+        max_buf_size,
+        read_waker: None,
+        write_waker: None,
     }
+}
```

Comment:
This provides a single object that is both the reader and writer, but in practice I think people will want those to be two different objects.

Reply:
I agree with the previous comment. In my case I would like to use these stream objects as drop-in replacements for `TcpStream` and related types.

A `TcpStream` is specific to one "node", i.e., the client has a `TcpStream`, and the server has a `TcpStream`. For testing code that uses a `TcpStream`, we can replace with `DuplexStream`s

```rust
let (client_rw, server_rw) = tokio::io::duplex(65_536);
```

Here we have both the client-side and server-side objects, and both are Read+Write. Thus, code written to generically handle IO on a `TcpStream` will function identically when replaced with the `DuplexStream`.

Now in my use case, each `TcpStream` is split into a reader (`OwnedReadHalf`) and a writer (`OwnedWriteHalf`), 

```rust
let (read_half, write_half) = tcp_stream.into_split();
```

and my code operates generically on the Read and on the Write objects. To replicate that, I would need the `SimplexStream` to work similarly:

```rust
let (client_read_half, server_write_half) = tokio::io::simplex(65_536);
```

And then you could stack two of these together to get functionality similar to the split of a `TcpStream`

```rust
let (client_read_half, server_write_half) = tokio::io::simplex(65_536);
let (server_read_half, client_write_half) = tokio::io::simplex(65_536);
```

Then I could pass `client_read_half` / `client_write_half` to my client-side code, and `stream_read_half` / `stream_write_half ` to my server-side code.

There is a bit of a question about the API. Do we want `tokio::io::simplex` to return a `SimplexStream` that is Read+Write, and then support `SimplexStream::into_split` that returns a `SimplexStreamReadHalf` and `SimplexStreamWriteHalf`? Or should `tokio::io::simplex` directly return the tuple `(SimplexStreamReadHalf, SimplexStreamWriteHalf)` as in my examples above?

(Also, I'm not sure if it makes sense to mix DuplexStreams and SimplexStreams or not, it sorta depends on how the simplex API looks.)

Reply:
IMO, we should not reinvent the wheel and try to reuse already existing APIs and facilities. It is true that for most of the use cases, we should rather return a pair of reader and writer instead of a single structure, that is a great point. Also, most of the channel-like structures already do this.
Furthermore, since `SimplexStream` already implements `AsyncRead` and `AsyncWrite`, we can use already existing [`split(…)`](https://docs.rs/tokio/latest/tokio/io/fn.split.html) to create reader and writer half (where possibly the split reader and writer half can be [`unsplit`](https://docs.rs/tokio/latest/tokio/io/struct.ReadHalf.html#method.unsplit)).

What do you think of it now?